ot/lp/dmmot.py

```diff
@@ -0,0 +1,319 @@
+        A dictionary containing the log of the discrete mmot problem:
+        - 'A': a dictionary that maps tuples of indices to the corresponding
+        primal variables. The tuples are the indices of the entries that are
+        set to their minimum value during the algorithm.
+        - 'primal objective': a float, the value of the objective function
+        evaluated at the solution.
+        - 'dual': a list of arrays, the dual variables corresponding to
+        the input arrays. The i-th element of the list is the dual variable
+        corresponding to the i-th dimension of the input arrays.
+        - 'dual objective': a float, the value of the dual objective function
+        evaluated at the solution.
+
+
+    References
+    ----------
+    .. [50] Ronak Mehta, Jeffery Kline, Vishnu Suresh Lokhande, Glenn Fung, &
+        Vikas Singh (2023). Efficient Discrete Multi Marginal Optimal
+        Transport Regularization. In The Eleventh International
+        Conference on Learning Representations.
+    .. [51] Jeffery Kline. Properties of the d-dimensional earth mover's
+        problem. Discrete Applied Mathematics, 265: 128-141, 2019.
+    .. [52] Leonid V Kantorovich. On the translocation of masses. Dokl. Akad.
+        Nauk SSSR, 37:227-229, 1942.
+
+    See Also
+    --------
+    ot.lp.discrete_mmot_converge : Minimized the d-Dimensional Earth Mover's
+    Distance (d-MMOT)
+    """
+
+    nx = get_backend(A)
+
+    # AA = [nx.copy(_) for _ in A]
+    AA = [nx.copy(A[:, j]) for j in range(A.shape[1])]
+
+    dims = tuple([len(_) for _ in AA])
+    xx = {}
+    dual = [nx.zeros(d) for d in dims]
+
+    idx = [0, ] * len(AA)
+    obj = 0
+
+    if verbose:
+        print('i minval oldidx\t\tobj\t\tvals')
+
+    while all([i < _ for _, i in zip(dims, idx)]):
+        vals = [v[i] for v, i in zip(AA, idx)]
+        minval = min(vals)
+        i = vals.index(minval)
+        xx[tuple(idx)] = minval
+        obj += (dist_monge(idx)) * minval
+        for v, j in zip(AA, idx):
+            v[j] -= minval
+        oldidx = nx.copy(idx)
+        idx[i] += 1
+        if idx[i] < dims[i]:
+            temp = dist_monge(idx) - dist_monge(oldidx) + dual[i][idx[i] - 1]
+            dual[i][idx[i]] += temp
+        if verbose:
+            print(i, minval, oldidx, obj, '\t', vals)
+
+    # the above terminates when any entry in idx equals the corresponding
+    # value in dims this leaves other dimensions incomplete; the remaining
+    # terms of the dual solution must be filled-in
+    for _, i in enumerate(idx):
+        try:
+            dual[_][i:] = dual[_][i]
+        except Exception:
+            pass
+
+    dualobj = nx.sum([nx.dot(A[:, i], arr) for i, arr in enumerate(dual)])
+
+    log_dict = {'A': xx,
+                'primal objective': obj,
+                'dual': dual,
+                'dual objective': dualobj}
+
+    if log:
+        return obj, log_dict
+    else:
+        return obj
+
+
+def discrete_mmot_converge(
```

Comment:
same here I would use `monge_mmot_1dgrid_optimize`to stet clearly what the function does. i also need more discussion about why one would optimize all distributions  together, you use is as some kind of "barycenter" since they all converge to a given distribution but be clear that it is not a barycenetr in the traditional OT sens.